### PR TITLE
Suggestion on mvcless app part Types.

### PR DESCRIPTION
**Pull request by jtkech**
Just as a suggestion.

The goal is to be sure we don't miss any Types. So, in place of starting with the exposed types of enabled features, here we start on the full list by using the public mvc `DefaultAssemblyPartDiscoveryProvider`. Then by excluding the types which are specific to features which are not enabled.

With the current implementation we have around 400 types. Here we have about 800 types.

Right now, from the app parts returned by the mvc static class, i use `ExportedTypes`. Maybe better to use `DefinedTypes` as mvc do. If we do this we get around 1700 types but still much less than 20000 types and here many types are shared by all tenants (through a shared app types list).

Quickly done, so need to be checked, maybe better to use `DefinedTypes` for app parts as mvc. For features excluded types, i think `ExportedTypes` are enough to inhib features which are not enabled.

Best.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Modules/ShellFeatureApplicationPart.cs [hunk 1]
```diff
@@ -70,5 +86,30 @@ public IEnumerable<string> GetReferencePaths()
...
+
+            lock (_synLock)
+            {
+                if (_applicationTypes != null)
+                {
+                    return _applicationTypes;
+                }
+                var hostingEnvironment = _httpContextAccessor.HttpContext.RequestServices.GetRequiredService<IHostingEnvironment>();
+
+                _applicationTypes = DefaultAssemblyPartDiscoveryProvider
```
Comment (sebastienros) [inline src/Microsoft.AspNetCore.Mvc.Modules/ShellFeatureApplicationPart.cs]: What is it supposed to find?

Comment (jtkech) [inline src/Microsoft.AspNetCore.Mvc.Modules/ShellFeatureApplicationPart.cs]: Find application parts, [see here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreServiceCollectionExtensions.cs#L77), so only those related to assemblies which have a reference on a primary mvc assembly, [see here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs#L18). Then we use `ExportedTypes` but maybe better to use `DefinedTypes`, [see here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs#L45). So, here we get types as if `AddMvcCore()` was called at the app level.

Then, after we exclude types exposed by features which are not enabled.

Not sure, just a suggestion, but the goal would be to be sure we don't miss any types.

If features Types don't take into account all the dependency graph, maybe better to 1st start from a more complete list, then exclude what is sufficient to inhibit features which are not enabled.

**Comment by sebastienros** [comment]
Sounds good then, instead of loading what we think is right, we load what MVC would and remove the features that are not enabled.
And is it correctly taking into account the types that are marked with `[OrchardFeatureDependency]`? I would assume if this is already handled in the code that gathers the DefinedTypes.

Would be good to test with a controller that is not enabled to ensure we can't reach it, then enable the feature and see we can. For both the default feature and a specific feature (with the attribute)

Also, last time we worked on a class to support precompiled views in this branch, and we added an AssemblyPart by ourselves, should we remove this code now? You know, where we were reusing the ICompileCache for razor.

**Comment by jtkech** [comment]
>... we load what MVC would and remove the features that are not enabled.

- Yes, this is the idea.

>... test if not enabled ... For both the default feature and a specific feature (with the attribute).

- Already tested with a default feature, i will do it for a specific one with the attribute.

>... we worked on a class to support precompiled views ... should we remove this code now?

- Oh yes, didn't remember this one, i will check it.

There is the difference between `DefinedTypes` and `ExportedTypes` but also the assemblies taken into account. So right now, just to have less types, from the mvc parts i still use `ExportedTypes`.

I'll change it to `DefinedTypes` as mvc do, we'll have more types (twice more) but we can test the impact.

Best.

**Comment by jtkech** [comment]
About precompiled views.

It doesn't work without our custom class because the related `ViewsFeatureProvider` (which i can see in the partManager FeatureProviders) is intended to act on an `AssemblyPart` related to the main app assembly.

Here, when we grab mvc parts i can see this assembly part that we use to resolve Types, but this is inside our `ShellFeatureApplicationPart` which doesn't inherit from `AssemblyPart` and whose name is different.

Will see if there is a simple solution.

**Comment by jtkech** [comment]
**About precompiled views and to replace SharedCompilerCacheProvider**

So, when adding mvc services a `DefaultCompilerCacheProvider` is added. This is through this service that `PopulateFeature()` of each views feature provider are called, [see here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultCompilerCacheProvider.cs#L24) and [see here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ApplicationPartManager.cs#L41). Then, a views feature provider will parse all `AssemblyPart`, [see here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs#L36).

So, we would need to use an `AssemblyPart` based on a fake `Assembly` with the same name as the main app... So, the simplest solution i found is to define and add to the part manager a views feature provider which do what we were doing in our custom `ICompilerCache`.

**Disadvantage**

- `ViewsFeatureProvider().PopulateFeature()` is called once but on each tenant. But i've seen that when called on another tenant it is still very fast (1 or 2ms). I think because the precompiled views assembly is already loaded. The 1st time it take 30ms as before.

- If there is no precompiled views, it is fast also the 1st time.

**Advatange**

- We don't limit the cache to only one `IApplicationFeatureProvider<ViewsFeature>`.

Let me know if it's ok or better to re-use our `SharedCompilerCacheProvider`.

Best.

**Comment by jtkech** [comment]
So, a simpler solution for precompiled views.

- Just by adding to part manager an `AssemblyPart` related to the main app assembly.

- Just the main assembly and not all its dependencies, so this `AssemblyPart` return only 4 more types.

- Then, because the name of this part is the name of the main app, precompiled views works.

- Then, because `GetReferencePaths()` of an assembly part use the dependency context (which can be used on the main assembly) to grab compilation libraries ([see here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs#L60)), we don't need anymore to implement `GetReferencePaths()` in our `ShellFeatureApplicationPart`.

Best.

**Comment by jtkech** [comment]
So, as a summary

- Not to miss any assemblies we 1st grab types as mvc do then we exclude those related to the shell.

- For assemblies grabbed from mvc we could use `ExportedTypes` (2 times less) but on another side mvc uses `DefinedTypes`. Let me know what is better to first try.

- Tests done with a controller that is not enabled to ensure we can't reach it. This with the default feature and a specific feature with the attribute.

- We add an `AssemblyPart` related to the app assembly, not its dependencies so only 4 more types.

- Then precompiled views works without our `SharedCompilerCacheProvider` and we don't need anymore to implement `GetReferencePaths()` in our `ShellFeatureApplicationPart`.

- Also, we don't limit the cache to only one `IApplicationFeatureProvider<ViewsFeature>`.

- Also completes a temporary fix, this to order features only once.

Best.

**Comment by sebastienros** [comment]
> precompiled views works without our SharedCompilerCacheProvider

It was also done so that two tenants don't compile the same view. You should try it but I assume it will break this, just go to the same view for two other tenants and look at the perf with and without it.

**Comment by jtkech** [comment]
You're right, this the disadvantage i mentioned in a previous comment.

**Disadvantage**

- `ViewsFeatureProvider().PopulateFeature()` is called once but on each tenant. But i've seen that when called on another tenant it is very fast (1 or 2ms). I think because the precompiled views assembly is already loaded. The 1st time it take 30ms as before.

- If there is no precompiled views, it is always fast.

**Comment by sebastienros** [comment]
You confirm that without precompiled views we are still caching compilation across many tenants?

**Comment by jtkech** [comment]
Yes,

But the `ViewsFeatureProvider` do its job on each tenant but because the precompiled views assembly is already loaded it is fast.

It take 30ms but only for the 1st tenant.

**Comment by sebastienros** [comment]
You are adding so much information in your answer that I am confused if you are answering the question. Repeating just in case: "without" precompiled views, are compiled views cached across tenants?

**Comment by jtkech** [comment]
Sorry, yes.

**Comment by sebastienros** [comment]
I don't trust you, looked at the code and there is no more CompilerCache maintained ;) which means each tenant gets its own, hence they can't be cached. Do you still have my console apps to test the startup perf for many tenants ?

**Comment by jtkech** [comment]
Oh sorry, i was saying that precompiled views works across tenants.

Oh, you're right about **"without" precompiled views**.

**Comment by jtkech** [comment]
I was too concentrated on precompiled views.

So, we need a shared cache across tenants wich do what the `DefaultCompilerCacheProvider` do [as here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultCompilerCacheProvider.cs#L23).

Will do it.

**Comment by sebastienros** [comment]
You don't really need to do anything, just undo the changes on SharedCompilerCacheProvider

**Comment by jtkech** [comment]
Ok, will do it.

But as i remember i've seen that this maybe we limit us to only one views feature provider. E.g if we add another one in partManager.Features. But not sure now, let me ckeck it.

**Comment by jtkech** [comment]
So, i reused a shared compiler cache.

But based on the `DefaultCompilerCacheProvider` to use all potential views feature providers in `ApplicationPartManager.FeatureProviders`.

But review to see if it is ok, it is as it seems to be done [here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultCompilerCacheProvider.cs#L24) and [here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ApplicationPartManager.cs#L41).

Then i tested that precompiled views works across tenants. And, without precompiled views, that views are cached across tenants ;)

**Comment by jtkech** [comment]
So, in my last commit.

- Exclude all feature types only once when grabbing application types which are cached.

- Then, when types are requested we just have to add shell features types.

- Not to expose types of disabled features, finally, for types grabbed from mvc i use `ExportedTypes` and filtered with `IsClass && !IsAbstract`, as we do for feature entries.

- Note: By doing this we still return more types because more assemblies are taken into account.

- If we have missing types by doing this, we would need to re-use `DefinedTypes` but this would mean that for feature entries we would also need to use `DefinedTypes` in place of `ExportedTypes`.

- So right now, when types are requested, we return 433 shell types + 227 application types.

Best.

**Comment by sebastienros** [comment]
What are the 227 application types? And what features are they assigned to?

**Comment by jtkech** [comment]
From core projects. I think there are not all necessary but ...

    {Orchard.DisplayManagement.PositionWrapper}
    {Orchard.DisplayManagement.ShapeAttribute}
    {Orchard.DisplayManagement.Views.CombinedResult}
    {Orchard.DisplayManagement.Views.ShapeResult}
    {Orchard.DisplayManagement.Views.ValueModel`1[T]}
    {Orchard.DisplayManagement.Title.PageTitleBuilder}
    {Orchard.DisplayManagement.Theming.ContentFileInfo}
    {Orchard.DisplayManagement.Theming.ThemeManager}
    {Orchard.DisplayManagement.Theming.ThemeSelectorResult}
    {Orchard.DisplayManagement.Theming.ThemingFileProvider}
    {Orchard.DisplayManagement.Shapes.BaseShape}
    {Orchard.DisplayManagement.Shapes.Composite}
    {Orchard.DisplayManagement.Shapes.Nil}
    {Orchard.DisplayManagement.Shapes.CoreShapes}
    {Orchard.DisplayManagement.Shapes.CoreShapesTableProvider}
    {Orchard.DisplayManagement.Shapes.DateTimeShapes}
    {Orchard.DisplayManagement.Shapes.Shape}
    ...

E.g maybe some of the above needed before we move shape tag helpers

    {Orchard.ContentManagement.ContentAliasManager}
    {Orchard.ContentManagement.ContentDefinitionManager}
    {Orchard.ContentManagement.DefaultContentManager}
    {Orchard.ContentManagement.DefaultContentManagerSession}
    {Orchard.ContentManagement.DefaultContentItemIdGenerator}
    {Orchard.ContentManagement.Records.ContentItemIndex}
    {Orchard.ContentManagement.Records.ContentItemIndexProvider}
    {Orchard.ContentManagement.Records.Migrations}
    {Orchard.ContentManagement.Handlers.UpdateContentsHandler}
    {Orchard.ContentManagement.Drivers.Coordinators.ContentPartHandlerCoordinator}
    {Orchard.ContentManagement.Cache.ContentDefinitionCacheContextProvider}
    {Orchard.ContentManagement.ContentElement}
    {Orchard.ContentManagement.ContentField}
    {Orchard.ContentManagement.ContentItem}
    {Orchard.ContentManagement.ContentItemConverter}
    {Orchard.ContentManagement.ContentItemMetadata}
    {Orchard.ContentManagement.ContentPart}
    ...

Some are related to features but defined in a core project (e.g Something.Abstractions)

    {Orchard.ContentTypes.Events.ContentFieldAttachedContext}
    {Orchard.ContentTypes.Events.ContentFieldDetachedContext}
    {Orchard.ContentTypes.Events.ContentPartAttachedContext}
    {Orchard.ContentTypes.Events.ContentPartContext}
    {Orchard.ContentTypes.Events.ContentPartCreatedContext}
    {Orchard.ContentTypes.Events.ContentPartDetachedContext}
    {Orchard.ContentTypes.Events.ContentPartFieldContext}
    ...
    {Orchard.Navigation.PagerParameters}
    {Orchard.Navigation.Pager}
    {Orchard.Navigation.PagerSlimParameters}
    {Orchard.Navigation.PagerSlim}
    {Orchard.Environment.Navigation.MenuItem}
    {Orchard.Environment.Navigation.NavigationBuilder}
    {Orchard.Environment.Navigation.NavigationHelper}
    {Orchard.Environment.Navigation.NavigationItemBuilder}
    {Orchard.Environment.Navigation.NavigationManager}
    ....

Best.

**Comment by sebastienros** [comment]
These types are unnecessary, this is also why we moved the TagHelpers in the modules, because we didn't want to scan the assemblies just for them. The types that are required are actually added by the modules, and associated to their features.

We stated that we would only discover the types from the modules, or register the ones from satellite assemblies if needed.

> The goal is to be sure we don't miss any Types.

I would only do that when we know we are missing some types, or we don't have a solution to add them. I am not aware of any issue right now that needs to be solved, or a scenario we want to support and requires this change.

**Comment by jtkech** [comment]
Ok, i understand.

With tag helpers we had `module => core project or any other package => mvc reference`.
So, i was thinking this scenario is still possible. But i agree, right now we can let things as it is.

Oh, not sure but i remember that in O1, in a razor view i was able to use e.g the contentManager. Will check if i can still use it with the current implementation.

So, i will revert this part but maybe good to keep some other parts e.g to complete a previous temporary workaround, this to order feature only once.

Best.

**Comment by jtkech** [comment]
So, i removed the usage of application types and i'm more confident with the current implementation.

So, because other things are not so useful i close this PR.

I'll open another PR for the little think related to features ordering.

Best.

**Pull request closed without merging**